### ISSUE TITLE
Marker drag event callback

### ISSUE THREAD

**Issue by chibatching** [googlemaps/android-maps-compose#10]
I want to get marker drag event callback and manipulate marker-related work.

So, `Marker` composable have to have dragging callback parameter like `onMarkerDrag` `onMarkerDragStart` `onMarkerDragEnd` similar with `onClick` and `onInfoWindow~`

It is an alternative to `OnMarkerDragListener` of Android view version.

**Comment by jpoehnelt** [comment]
@chibatching Thank you for opening this issue. 🙏
Please check out these other resources that might be applicable:

- [Support Console](https://cloud.google.com/support/) - If you have a support contract.
- [Discord](https://discord.gg/9UpV862xZ9) - chat with other developers
- [StackOverflow](http://stackoverflow.com/questions/tagged/google-maps) - use the `google-maps` tag
- [Google Maps Platform issue trackers](https://developers.google.com/maps/support/#issue_tracker)


_This is an automated message, feel free to ignore._

**Comment by chibatching** [comment]
@arriolac
I have to know marker LatLng after drag to execute some process using the position after the move.
It is not enough to observe only the drag state.

My usecase is like below

```kotlin
var markerPoint by remember { mutableStateOf(defaultLatLng) }

GoogleMap() {
    Marker(
        position = markerPoint,
        draggable = true,
        onMarkerDragEnd = { marker ->
            markerPoint = maker.position
            viewModel.updateUserSelectedPosition(marker.position)
        },
    )
}

**Comment by chibatching** [comment]
I think it doesn't need to be a callback style if the drag state has marker object or position info.

**Comment by arriolac** [comment]
> I think it doesn't need to be a callback style if the drag state has marker object or position info.

Yeah, this is what I'm thinking as well. There should be a `Marker` property in the `MarkerDragState` object. Will add that.

**Comment by chibatching** [comment]
I updated my PR #11 to add `Marker` property.
Please check it.

**Comment by arriolac** [comment]
Please see my comment on https://github.com/googlemaps/android-maps-compose/pull/11#issuecomment-1035586376 for how this feature should be approached.

**Comment by bubenheimer** [comment]
AFAICT the implementation here does not take into account the difference between drag state and drag events. The `onMarkerDrag...` callbacks reflect events, while the current 'MarkerState' approach reflects state. This implies information loss; some or all drag events may never be seen by an observer. An approach like this may be reasonable when there is only a single type of event, DRAG, but there are 3 different types, and an observer may never notice that the marker was dragged, as START, and DRAG, may fall through the cracks; even a distinct END may not be seen by a slow observer when another drag event starts quickly. Really, the only thing I can observe here is whether I am currently dragging or not.

For example, I have a use case where a user's marker drag events are not committed to the data model when the marker is dragged to an invalid position. This looks hard to implement reliably at the moment, if possible at all. I don't see a way to access the raw events via a listener, either.

**Comment by bubenheimer** [comment]
There also seems to be a misconception about what state hoisting means. When I hoist state, I control the state at a higher level of the composition, and the state object is not present at deeper levels of the composition, which cannot directly change the state. Yet here I am expected to pass the "hoisted" state object into the composition, and it will change under my nose when the user drags the marker. That is the exact opposite of state hoisting. It means a lot of trouble for my outside data model.

**Comment by arriolac** [comment]
@bubenheimer thanks for the feedback on the current marker drag API. I'm a bit confused what you mean by information loss as all marker drag transition states should be reported in `MarkerState.dragState`—maybe I'm missing something? 

From your feedback, seems like it is desirable to have an event-based API to accomplish validation type actions as the current state-based approach for drag state is difficult to work with. If you can capture some suggestions you have on a separate issue we can reevaluate and improve upon the existing API.

**Comment by bubenheimer** [comment]
Thanks. In general there is no way to guarantee that an observer will see transitory states of a MutableState, the observer only gets to see snapshot state (not events), and it may not even see all of the snapshot states. This is a fundamental characteristic of snapshot state. For background, see the bottom of the `snapshotFlow` documentation: https://developer.android.com/reference/kotlin/androidx/compose/runtime/package-summary#snapshotFlow(kotlin.Function0)

I will propose something in a separate issue.

Regarding "hoisting": I suggest simply removing mention of that term from the MarkerState docu, as it has a well-defined and different meaning in the Compose world; it's not just about moving a state object up in the composition, the state object is eliminated from lower levels of the composition and they cannot directly modify it; that's essential: https://developer.android.com/jetpack/compose/state#state-hoisting

**Comment by zaheeroz** [comment]
Hope we will see this event soon, both drag start and drag end. I could do with drag start but drag end is critical.

**Comment by bubenheimer** [comment]
@zaheeroz You may be able to "reconstruct" drag start & end to some degree from the information that is available, but you need to take into account that you may miss some ends and/or starts, when a new start-end cycle starts too quickly after the last one ends.

**Comment by zaheeroz** [comment]
Thanks @bubenheimer that's great. Could you provide a link that can help in this regard.  I can start a drag by first clicking on the marker to register a 'drag start'  then drag the marker. But the problem is even click fires on 'pressed' and not ' on release' so I cant get the location where the marker was dragged (end of drag).  So any help in this regard will be greatly appreciated.

**Comment by bubenheimer** [comment]
@zaheeroz I don't have a link, sorry.

**Comment by wangela** [comment]
I believe the initial request here was addressed in v2.0, so closing this issue. Conversation continues about improving marker drag state in #149.
